src/SimpleNameRegister.sol

```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+///@title An on-chain name registry
+///@author Calnix
+contract SimpleNameRegister {
+    
+    /// @notice map a name to an address to identify current holder 
+    mapping (string => address) public holder;    
+
+    /// @notice emit an event when a name is registered or released
+    event Register(address indexed holder, string name);
+    event Release(address indexed holder, string name);
+
+    /// @notice user can register an available name
+    function register(string memory name) public {
+        require(holder[name] == address(0), "Already registered!");
+        holder[name] = msg.sender;
+        emit Register(msg.sender, name);
+    }
+
+    /// @notice holder can release a name, making it available
+    function release(string memory name) public {
```

Comment:
we aren't going to get into gas optimization until issue 3 or 4, but you should lookup [calldata](https://docs.soliditylang.org/en/v0.8.13/types.html?highlight=calldata#data-location)

Reply:
Could you please confirm if my understanding is correct?

When using `(string memory name)`, a copy of the parameter is passed into  `delete holder[name] `, as per the function below:
```
    function release(string memory name) public {
        require(holder[name] == msg.sender, "Not your name!");
        delete holder[name];
        emit Release(msg.sender, name);
    }
}
```
However, using `(string calldata name)`, the parameter would be passed directly into `delete holder[name] `, without making a copy, thereby saving gas. 
```
    function release(string calldata name) public {
        require(holder[name] == msg.sender, "Not your name!");
        delete holder[name];
        emit Release(msg.sender, name);
    }
}
```

Reply:
Yeah that's pretty much it.  There is already a location for the calldata that can be referenced. So as long as you just need it for readOnly, no point in making a fresh copy in a new memory location (which is what setting it to 'memory') does.

One gotcha on this is that if you pass in a string (or bytes) from a different internal function where you had just created that string in memory, then you can't use calldata here.  Example:

```
contract X {
 function sendString() internal {
    string memory y = "Raja is cool"l
    callFunction(y);
 }

 function callFunction(bytes calldata y) {}  // ERROR -- because calldata is only used from external calls and y was created in memory before
}
```